tests/jax/core_system_components/launcher_test.py

```diff
@@ -15,4 +15,381 @@
+        return "test_parameter_server_second_mock"
+
+    return parameter_server_second
+
+
+def test_initiator_multi_process() -> None:
+    """Test the constructor of Launcher in the case of multi process"""
+    launcher = Launcher(multi_process=True)
+
+    assert launcher._multi_process == True
+    assert launcher._name == "System"
+    assert launcher._sp_trainer_period == 10
+    assert launcher._sp_evaluator_period == 10
+    assert launcher._terminal == "current_terminal"
+
+    assert launcher._nodes_on_gpu == []
+    assert isinstance(launcher._program, lp.Program)
+    assert launcher._program._name == "System"
+    assert launcher._program._groups == {}
+    assert launcher._program._current_group == None
+
+    assert not hasattr(launcher, "_nodes")
+    assert not hasattr(launcher, "_node_dict")
+
+
+def test_initiator_non_multi_process() -> None:
+    """Test the constructor of Launcher in the case of one process"""
+    launcher = Launcher(multi_process=False)
+
+    assert launcher._multi_process == False
+    assert launcher._name == "System"
+    assert launcher._sp_trainer_period == 10
+    assert launcher._sp_evaluator_period == 10
+    assert launcher._terminal == "current_terminal"
+
+    assert launcher._nodes == []
+    assert launcher._node_dict == {
+        "data_server": None,
+        "parameter_server": None,
+        "executor": None,
+        "evaluator": None,
+        "trainer": None,
+    }
+
+    assert not hasattr(launcher, "_nodes_on_gpu")
+    assert not hasattr(launcher, "_program")
+
+
+def test_add_multi_process(mock_data_server_fn: Callable) -> None:
+    """Test add method in the Launcher for the case of multi process
+
+    Args:
+        mock_data_server_fn
+    """
+    launcher = Launcher(multi_process=True)
+    data_server = launcher.add(
+        mock_data_server_fn,
+        node_type=NodeType.reverb,
+        name="data_server_test",
+    )
+
+    assert list(launcher._program._groups.keys()) == ["data_server_test"]
+    assert (
+        launcher._program._groups["data_server_test"][-1]._priority_tables_fn
+        == mock_data_server_fn
+    )
+    assert [data_server] == launcher._program._groups["data_server_test"][
+        -1
+    ]._created_handles
+
+    assert not hasattr(launcher, "_nodes")
+    assert not hasattr(launcher, "_node_dict")
+
+
+def test_add_multi_process_two_add_calls(
+    mock_data_server_fn: Callable, mock_parameter_server_fn: Callable
+) -> None:
+    """Test calling add more than one time method in the Launcher for the case of multi process
+
+    Args:
+        mock_data_server_fn
+        mock_parameter_server_fn
+    """
+    launcher = Launcher(multi_process=True)
+    data_server = launcher.add(
+        mock_data_server_fn,
+        node_type=NodeType.reverb,
+        name="data_server_test",
+    )
+    parameter_server = launcher.add(
+        mock_parameter_server_fn,
+        node_type=NodeType.corrier,
```

Comment:
This was changed to `NodeType.courier` by Asad in another PR, so this code should not work.